### ISSUE TITLE
Don't sort output of OpenAPI 3 emitter by default

### ISSUE THREAD

**Issue by mikekistler** [microsoft/typespec#1618]
PR #1611 introduced sorting unconditionally to the OpenAPI 3 output. But sorted output is undesirable in some situations so this feature should be configurable with an emitter option.  And I would argue that we should default to not-sorted, as it maps better from Cadl source to emitted output.

**Comment by nguerrera** [comment]
In my opinion, this is not as simple as putting #1611 behind a flag. The behavior before #1611 was not deliberately designed, nor documented, nor tested. And it was not particularly useful either IMHO.

Here's an example in the playground on a commit before #1611:

[Plaground Link](https://cadlplayground.z22.web.core.windows.net/prs/1610/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKdXNpbmcgQ2FkbC5IdHRwOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCiAgdmVyc2lvbjogIjEuMC4wIiwKfSkKbmFtZXNwYWNlIERlbW%2FHLzsKCgpAZXJyb3IKbW9kZWwgRcQMIMRgY29kZTogaW50MzI7CiAgbWVzc2FnZTogc3RyaW5nOwp9Cscz5wCBxDRpZMogICB3ZWlnaHTLRGNvbG9yOiAicmVkIiB8ICJibHVlIsVKCkByb3V0ZSgiL3fFTnMiKQpAdGFnKOcA38QQaW50ZXJm5ADBxxTFc0DEfHJlYWQoQHBhdGjrAIMpOscpICB85gDYxGs%3D)

Why does `Widget` come before `Error` in the output schemas? I can only answer this by digging into the details of the code.

I think it's actually a fair deal of work to specify source order of all entities in a way that user's would understand and find useful as the default output for OpenAPI. There are two levels of issues here: 

1) The order in the compiler's type graph can be surprising. For example:
![image](https://user-images.githubusercontent.com/75470/217383900-27550eaf-5771-4ff1-b168-d6c9a6fbe5c6.png)
![image](https://user-images.githubusercontent.com/75470/217383893-3ab3d31f-e388-4b7e-874a-46b7fbb24d32.png)
![image](https://user-images.githubusercontent.com/75470/217383919-1be6d3ce-2b01-4c94-aa90-1fc0c306da59.png)
 
2) OpenAPI emit order was not the same as type graph order before #1611 either. It was further sensitive to its own choices in how it traversed things.

Without doing that work, I would argue that #1611 is a net improvement and a better default than existed previously. It also matches what cadl-autorest has done for ages.

Also consider that we do not sort *everything* with #1611, we only sort #/paths, #/components/parameters, and #/components/schemas by their keys. I'm not aware of any real scenarios where the order of these matters. Are there any in practice? I know that doing things like re-ordering properties would be very problematic, but we don't do that and thankfully property order is well and intuively defined in the compiler's type graph.

**Comment by bterlson** [comment]
FWIW I think the type graph order is well defined in general (post-order traversal of reference graph starting from types found in the main file and imported files in the order they are imported), but considering the previous order was in fact some subtle difference from this, your point still stands.

**Comment by mikekistler** [comment]
Hmmm ... you make a good point Nick. Maybe the prior behavior is less well-defined than I imagined.

In any event this is not something we would undertake in Zinc given all the other work on our plate, so let's just put this in the backlog and revisit at the beginning of Ga.

**Comment by nguerrera** [comment]
The type graph order is well-defined in the sense that it is deterministic and we could write a specification that the current implementation satisfies, but I recall that  the last time we talked about ordering of members in a namespace, we wanted to say that it was an implementation detail subject to change. I suspect that if we had to promise the order, we would want to make some changes to it first.

However, even if the current type graph order is written in stone, I'm not convinced that it matches what a user would want in their OpenAPI output. (Assuming the OpenAPI emitter preserved it, which it didn't, but that's a separate point.) If someone is asking for the OpenAPI order to match their source, I don't think they are expecting that adding a reference can move a type around. 

> (post-order traversal of reference graph starting from types found in the main file and imported files in the order they are imported)

I think there's more subtletly here. For example, the way that I read that, I would not expect this:

<img width="469" alt="image" src="https://user-images.githubusercontent.com/75470/217601792-8a36e1f0-1dae-4b41-81b1-b571c6e27dea.png">

References "inside" a model don't bring forward types the way references "outside" do, it seems. I think this is the relevant code:

https://github.com/microsoft/cadl/blob/0a4c49db7014eec73840d6a85f89dc9616fdf493/packages/compiler/core/checker.ts#L2316-L2323

There is also the question of how reference edges coming out of the same node are ordered. By inspecting nearby code, I found this peculiarity:

![isextendsorder](https://user-images.githubusercontent.com/75470/217607634-100791d0-6b69-4e86-bce6-ef3fd8a3c91f.gif)
